### PR TITLE
Make AzureFunctionAppContainerV1 compatible with Centauri

### PR DESCRIPTION
**Pull request by patelchandni**
**Task name**: AzureFunctionAppContainerV1 

**Description**: Skipping couple of steps for function apps on Centauri because Centauri does not support:
1. Kudu/SCM site
2. Restart API
3. Slots

**Documentation changes required:** (Y/N) NO

**Added unit tests:** (Y/N) NO

**Attached related issue:** (Y/N) NO

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

### CODE REVIEW

**Comment by patelchandni** [comment]
waiting for this PR to be released first: https://github.com/microsoft/azure-pipelines-tasks/pull/18198

**Comment by starkmsu** [comment]
This must be updated on the last version from master - a lot of changes were rolled back there

File: Tasks/AzureFunctionAppContainerV1/azurefunctiononcontainerprovider.ts [hunk 1]
```diff
@@ -37,38 +40,55 @@ export class AzureFunctionOnContainerDeploymentProvider{
...
+        }
     }
 
     public async UpdateDeploymentStatus(isDeploymentSuccess: boolean) {
-        if(this.kuduServiceUtility) {
-            await addReleaseAnnotation(this.azureEndpoint, this.appService, isDeploymentSuccess);
-            this.activeDeploymentID = await this.kuduServiceUtility.updateDeploymentStatus(isDeploymentSuccess, null, {'type': 'Deployment', slotName: this.appService.getSlot()});
-            tl.debug('Active DeploymentId :'+ this.activeDeploymentID);
-        }
+        if (!this.isCentauri){
```
Comment (onetocny) [inline Tasks/AzureFunctionAppContainerV1/azurefunctiononcontainerprovider.ts:82]: You can merge these two if conditions to simplify code:

```
if (!this.isCentauri && this.kuduServiceUtility)
```

File: Tasks/AzureFunctionAppContainerV1/azurefunctiononcontainerprovider.ts [hunk 2]
```diff
@@ -37,38 +40,55 @@ export class AzureFunctionOnContainerDeploymentProvider{
...
-        let linuxFunctionStorageSetting: string = ''; 
-        if (!this.taskParams.AppSettings || this.taskParams.AppSettings.indexOf(linuxFunctionStorageSettingName) < 0) { 
-            linuxFunctionStorageSetting = `${linuxFunctionStorageSettingName} ${linuxFunctionStorageSettingValue}`; 
-        }
-        this.taskParams.AppSettings = this.taskParams.AppSettings ? this.taskParams.AppSettings.trim() + " " + linuxFunctionStorageSetting : linuxFunctionStorageSetting;
-        let customApplicationSettings = ParameterParser.parse(this.taskParams.AppSettings);
-        await this.appServiceUtility.updateAndMonitorAppSettings(customApplicationSettings);
         
-        await this.appServiceUtility.updateScmTypeAndConfigurationDetails();
+        if (this.isCentauri){
```
Comment (onetocny) [inline Tasks/AzureFunctionAppContainerV1/azurefunctiononcontainerprovider.ts:58]: There is a lot of code duplication in following lines and it is really hard to determine what are the differences between those two if branches. Can you please try to simplify it? For example:

```ts
await containerDeploymentUtility.deployWebAppImage(this.taskParams, !this.isCentauri);
let linuxFunctionStorageSetting: string = ''; 
if (!this.taskParams.AppSettings || this.taskParams.AppSettings.indexOf(linuxFunctionStorageSettingName) < 0) { 
    linuxFunctionStorageSetting = `${linuxFunctionStorageSettingName} ${linuxFunctionStorageSettingValue}`; 
}
this.taskParams.AppSettings = this.taskParams.AppSettings ? this.taskParams.AppSettings.trim() + " " + linuxFunctionStorageSetting : linuxFunctionStorageSetting;
let customApplicationSettings = ParameterParser.parse(this.taskParams.AppSettings);

if (this.isCentauri){
 await this.appServiceUtilityExt.updateAndMonitorAppSettingsCentauri(customApplicationSettings, fal
} else {
 await this.appServiceUtility.updateAndMonitorAppSettings(customApplicationSettings);
 await this.appServiceUtility.updateScmTypeAndConfigurationDetails();            
}
```

File: Tasks/AzureFunctionAppContainerV1/operations/AzureAppServiceUtilityExt.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        else {
+            console.log(tl.loc('AppServiceApplicationSettingsAlreadyPresent'));
+        }
+
+        if (perSlot){
+            await this._appService.patchApplicationSettingsSlot(addProperties);
+        }        
+        return isNewValueUpdated;
+    }
```
Comment (onetocny) [inline Tasks/AzureFunctionAppContainerV1/operations/AzureAppServiceUtilityExt.ts:73]: This is exact copy of `AzureAppServiceUtility.updateAndMonitorAppSettings` from `arm-rest-v2`. The only difference is the `if (perSlot)` condition here (which is always `false` in you case btw). My suggestion here is to add `perSlot` parameter to original `updateAndMonitorAppSettings` in `arm-rest-v2` and similar condition there.

**Review by onetocny — CHANGES_REQUESTED** [review]
Thank you for your PR. However there is quite a lot of code duplication which leads to really complicated code, poor readability and increases future efforts for maintenance. We should overcome code repetition as much as possible to keep the code clean and elegant. Please see more details [here](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

File: Tasks/AzureFunctionAppContainerV1/operations/AzureAppServiceUtilityExt.ts (hunk 3 above)
Comment (patelchandni) [reply Tasks/AzureFunctionAppContainerV1/operations/AzureAppServiceUtilityExt.ts:73]: Sure, this task is not using arm-rest-v2, but I will create a PR against https://github.com/microsoft/azure-pipelines-tasks-common-packages

Comment (patelchandni) [reply Tasks/AzureFunctionAppContainerV1/operations/AzureAppServiceUtilityExt.ts:73]: PR created: https://github.com/microsoft/azure-pipelines-tasks-common-packages/pull/148

**Review by onetocny — APPROVED** [review]

**Review by starkmsu — APPROVED** [review]

**Pull request merged**